File: scripts/pgo-train.mjs
import { readFileSync } from 'node:fs';
import { createRequire } from 'node:module';
import { dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { glob } from 'glob';

const here = dirname(fileURLToPath(import.meta.url));
const repoRoot = resolve(here, '..');
const require = createRequire(resolve(repoRoot, 'packages/svelte-rs2/compiler/index.js'));
const native = require(resolve(repoRoot, 'packages/svelte-rs2/compiler/native/binding.cjs'));

const rounds = Number(process.env.PGO_ROUNDS ?? 3);
const corpusDirs = [
    resolve(repoRoot, 'tasks/benchmark/benches/compiler'),
    resolve(repoRoot, 'tasks/compiler_tests/cases2'),
    ...(process.env.PGO_CORPUS ?? '').split(/\s+/).filter(Boolean),
];

const files = [];
for (const dir of corpusDirs) {
    for (const rel of await glob('**/*.{svelte,svelte.js,svelte.ts}', { cwd: dir })) {
        const path = resolve(dir, rel);
        try {
            files.push([path, readFileSync(path, 'utf8')]);
        } catch {
            /* unreadable entry — skip */
        }
    }
}

process.stderr.write(`pgo-train: ${files.length} files x ${rounds} rounds\n`);

const modes = [];
for (const generate of ['client', 'server']) {
    for (const dev of [false, true]) {
        modes.push({ generate, dev });
    }
}

for (let round = 0; round < rounds; round++) {
    for (const [filename, source] of files) {
        const isModule = filename.endsWith('.js') || filename.endsWith('.ts');
        for (const mode of modes) {
            try {
                if (isModule) {
                    native.compileModule(source, { ...mode, filename });
                } else {
                    native.compile(source, { ...mode, filename });
                }
            } catch {
                /* invalid fixtures are expected in the corpus */
            }
        }
    }
}

File: tasks/benchmark/src/bin/pgo_train.rs
use std::env;
use std::fs;
use std::hint::black_box;

use benchmark as _;
use glob::glob;

const DEFAULT_ROUNDS: u32 = 3;

fn is_module(path: &str) -> bool {
    path.ends_with(".svelte.js") || path.ends_with(".svelte.ts")
}

fn corpus_patterns(extra: &[String]) -> Vec<String> {
    let root = concat!(env!("CARGO_MANIFEST_DIR"), "/../..");
    let mut patterns = vec![
        format!("{root}/tasks/benchmark/benches/compiler/**/*.svelte"),
        format!("{root}/tasks/benchmark/benches/compiler/**/*.svelte.js"),
        format!("{root}/tasks/compiler_tests/cases2/**/*.svelte"),
        format!("{root}/tasks/compiler_tests/cases2/**/*.svelte.js"),
    ];
    for dir in extra {
        patterns.push(format!("{dir}/**/*.svelte"));
        patterns.push(format!("{dir}/**/*.svelte.js"));
    }
    patterns
}

fn main() {
    let mut rounds = DEFAULT_ROUNDS;
    let mut extra_dirs: Vec<String> = Vec::new();
    for arg in env::args().skip(1) {
        match arg.parse::<u32>() {
            Ok(value) => rounds = value,
            Err(_) => extra_dirs.push(arg),
        }
    }

    let mut sources: Vec<(String, String, bool)> = Vec::new();
    for pattern in corpus_patterns(&extra_dirs) {
        let Ok(entries) = glob(&pattern) else {
            continue;
        };
        for entry in entries.filter_map(Result::ok) {
            let path = entry.display().to_string();
            let Ok(source) = fs::read_to_string(&entry) else {
                continue;
            };
            let module = is_module(&path);
            sources.push((path, source, module));
        }
    }
    sources.sort();
    sources.dedup_by(|a, b| a.0 == b.0);

    eprintln!("pgo_train: {} files x {rounds} rounds", sources.len());

    let generate_modes = [
        svelte_compiler::GenerateMode::Client,
        svelte_compiler::GenerateMode::Server,
    ];

    for _ in 0..rounds {
        for (path, source, module) in &sources {
            for dev in [false, true] {
                for generate in generate_modes {
                    if *module {
                        let options = svelte_compiler::ModuleCompileOptions {
                            dev,
                            generate,
                            filename: path.clone(),
                            ..svelte_compiler::ModuleCompileOptions::default()
                        };
                        black_box(svelte_compiler::compile_module(source, &options));
                    } else {
                        let options = svelte_compiler::CompileOptions {
                            dev,
                            generate,
                            filename: path.clone(),
                            ..svelte_compiler::CompileOptions::default()
                        };
                        black_box(svelte_compiler::compile(source, &options));
                    }
                }
            }
        }
    }
}
